File: Code Samples/Stack - Nikita Choudhari/Stack Implemenation using Linked List/stack.cpp
#include <iostream>

struct Node {
    int data;
    Node* next;
};

class Stack {
private:
    Node* top;

public:
    Stack() {
        top = nullptr;
    }

    void push(int data) {
        Node* newNode = new Node;
        newNode->data = data;
        newNode->next = top;
        top = newNode;
    }

    void pop() {
        if (top == nullptr) {
            std::cout << "Stack Underflow\n";
        } else {
            Node* temp = top;
            top = top->next;
            delete temp;
        }
    }

    int peek() {
        if (top == nullptr) {
            std::cout << "Stack is Empty\n";
            return -1;
        }
        return top->data;
    }

    bool isEmpty() {
        return top == nullptr;
    }
};

int main() {
    Stack stack;
    stack.push(10);
    stack.push(20);
    stack.push(30);
    std::cout << stack.peek() << "\n";
    stack.pop();
    std::cout << stack.peek() << "\n";
    return 0;
}

File: Code Samples/Stack - Nikita Choudhari/Stack Implemenation using Linked List/stack.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Stack:
    def __init__(self):
        self.top = None

    def push(self, data):
        new_node = Node(data)
        new_node.next = self.top
        self.top = new_node

    def pop(self):
        if self.top is None:
            print("Stack Underflow")
        else:
            self.top = self.top.next

    def peek(self):
        if self.top is None:
            print("Stack is Empty")
            return -1
        return self.top.data

    def is_empty(self):
        return self.top is None


stack = Stack()
stack.push(10)
stack.push(20)
stack.push(30)
print(stack.peek())
stack.pop()
print(stack.peek())
